tests/python_package_test/test_sklearn.py

```diff
@@ -617,6 +617,104 @@ def test_nan_handle(self):
+                  'seed': 123}
+        params_fit = {'X': X_train,
+                      'y': y_train,
+                      'early_stopping_rounds': 5,
+                      'verbose': False}
+
+        def fit_and_check(eval_set_names, metric_names, assumed_iteration, first_metric_only):
+            params['first_metric_only'] = first_metric_only
+            gbm = lgb.LGBMRegressor(**params).fit(**params_fit)
+            self.assertEqual(len(gbm.evals_result_), len(eval_set_names))
+            for eval_set_name in eval_set_names:
+                self.assertIn(eval_set_name, gbm.evals_result_)
+                self.assertEqual(len(gbm.evals_result_[eval_set_name]), len(metric_names))
+                for metric_name in metric_names:
+                    self.assertIn(metric_name, gbm.evals_result_[eval_set_name])
+
+                    actual = len(gbm.evals_result_[eval_set_name][metric_name])
+                    expected = assumed_iteration + (params_fit['early_stopping_rounds']
+                                                    if eval_set_name != 'training' else 0)
+                    self.assertEqual(actual, expected)
+                    #self.assertEqual(assumed_iteration if eval_set_name != 'training' else 0, gbm.best_iteration_)
+                    self.assertEqual(assumed_iteration, gbm.best_iteration_)
+
+        # training data as eval_set
+        params_fit['eval_set'] = (X_train, y_train)
+        fit_and_check(['training'], ['l2'], 30, False)
+        fit_and_check(['training'], ['l2'], 30, True)
+
+        # # single eval_set
+        iter_1 = 2
+        iter_2 = 8
+        params_fit['eval_set'] = (X_test, y_test)
+        fit_and_check(['valid_0'], ['l2'], iter_1, False)
+        fit_and_check(['valid_0'], ['l2'], iter_1, True)
+
+        params_fit['eval_metric'] = "l2"
+        fit_and_check(['valid_0'], ['l2'], iter_1, False)
+        fit_and_check(['valid_0'], ['l2'], iter_1, True)
+
+        params_fit['eval_metric'] = "l1"
+        fit_and_check(['valid_0'], ['l1', 'l2'], iter_1, False)
+        fit_and_check(['valid_0'], ['l1', 'l2'], iter_2, True)
+
+        params_fit['eval_metric'] = ["l1", "l2"]
+        fit_and_check(['valid_0'], ['l1', 'l2'], iter_1, False)
+        fit_and_check(['valid_0'], ['l1', 'l2'], iter_2, True)
+
+        params_fit['eval_metric'] = ["l2", "l1"]
+        fit_and_check(['valid_0'], ['l1', 'l2'], iter_1, False)
+        fit_and_check(['valid_0'], ['l1', 'l2'], iter_1, True)
+
+        params_fit['eval_metric'] = ["l2", "regression", "mse"]  # test aliases
+        fit_and_check(['valid_0'], ['l2'], iter_1, False)
+        fit_and_check(['valid_0'], ['l2'], iter_1, True)
+
+        # two eval_set
+        X_test1, X_test2, y_test1, y_test2 = train_test_split(X_test, y_test, test_size=0.3, random_state=72)
+        params['learning_rate'] = 0.4
+
+        iter_l2_val0 = 6
+        iter_l2_val1 = 23
+        iter_l1_val0 = 4
+        iter_l1_val1 = 23
```

Comment:
`iter_l2_val1 = 23` and `iter_l1_val1 = 23`: the same values for different metrics make no sense for tests.